src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileFunction.java

```diff
@@ -97,31 +99,50 @@ public SkyValue compute(SkyKey skyKey, Environment env)
   }
 
   public static BazelLockFileValue getLockfileValue(RootedPath lockfilePath) throws IOException {
-    BazelLockFileValue bazelLockFileValue;
     try {
-      String json = FileSystemUtils.readContent(lockfilePath.asPath(), UTF_8);
-      Matcher matcher = LOCKFILE_VERSION_PATTERN.matcher(json);
-      int version = matcher.find() ? Integer.parseInt(matcher.group(1)) : -1;
-      if (version == BazelLockFileValue.LOCK_FILE_VERSION) {
-        bazelLockFileValue =
-            GsonTypeAdapterUtil.createLockFileGson(
-                    lockfilePath
-                        .asPath()
-                        .getParentDirectory()
-                        .getRelative(LabelConstants.MODULE_DOT_BAZEL_FILE_NAME))
-                .fromJson(json, BazelLockFileValue.class);
-      } else {
-        // This is an old version, needs to be updated
-        // Keep old version to recognize the problem in error mode
-        bazelLockFileValue = EMPTY_LOCKFILE.toBuilder().setLockFileVersion(version).build();
+      Gson gson =
+          GsonTypeAdapterUtil.createLockFileGson(
+              lockfilePath
+                  .asPath()
+                  .getParentDirectory()
+                  .getRelative(LabelConstants.MODULE_DOT_BAZEL_FILE_NAME));
+      int version;
+      try (var reader =
+          new BufferedReader(
+              new InputStreamReader(lockfilePath.asPath().getInputStream(), UTF_8))) {
+        BazelLockFileValue bazelLockFileValue = gson.fromJson(reader, BazelLockFileValue.class);
```

Comment:
We did consider this but we wanted to avoid parsing the whole file if the version is already old. Which I think is better.

Reply:
Could you explain what worries you about parsing the whole file if the version is old? If it is about performance, I would find optimizing for the common case of a matching version more natural. 

But then again, I don't know whether lockfiles will get so large that reading them into memory instead of streaming them makes a measurable difference even for large monorepos, so it's possible that there isn't much impact either way.

Reply:
Performance & why parse the whole file if we might not use all these info?
+ If the version is old we will probably get a parsing exception then parse it again to get the version. So I think just getting the version first and parse it once is better.

Reply:
Yes, this change definitely does more work in the case where the version doesn't match. But avoiding that work, as is done at HEAD, also comes at a cost: The lockfile is fully read into memory (instead of streaming it into the JSON parser) and searched for the regex every time, which is overhead that would be avoided with this change.

I would say it's a tradeoff and lean towards optimizing the common case, but it's quite possible that it won't make much of a difference. I can try to benchmark this, but lack access to a realistic very large MODULE.bazel file.

Reply:
to fully lay it out:

- HEAD: always stores whole lockfile as a string in memory, in addition to the parsed lockfile structures; parses JSON once in happy path, 0 times otherwise
- this PR: only stores the parsed lockfile structures; parses JSON once in happy path, twice otherwise

I wonder if we can try to "cheat" a bit and take the best of both worlds: always read the first 5 lines (or 100 characters or whatever) of the lockfile to look for the version. If happy, parse JSON by streaming, otherwise directly barf. This avoids both the memory issue and the double-parsing, and the downside of "we might have a lockfile whose version is specified lower down in the file" is negligible IMO.